test/tests.bats

```diff
@@ -52,10 +53,7 @@ setup_file() {
 
   # this relies on the checksum file being updated after all changes have been applied
   wait_until_change_detection_event_completes mail
-
-  # wait for ClamAV to be fully setup or we will get errors on the log
-  repeat_in_container_until_success_or_timeout 60 mail test -e /var/run/clamav/clamd.ctl
-
+  wait_for_service mail postfix
```

Comment:
There is two accounts added a few lines up. I am not sure if they are the cause of the problem here, if they introduce a change that gets detected and handled while another is in process.

Probably need to check how many changes are being detected / processed.

---

Technically we can use the `repeat_in_container_until_success_or_timeout` as  a workaround for each `nc` command below. Then if they fail due, they've got ample time to try again until everything should be back up to be successful.

Reply:
> Probably need to check how many changes are being detected / processed.

Confirmed. 2 changes detected.

<details>
<summary>Investigation</summary>

Interestingly, this is due to the `docker run` of a separate instance of the DMS image that uses an alternative approach to append the account to the `postifix-accounts.cf`. And it's this specific approach that is consistently producing two changes (_I've added sufficient sleep before and after, checking the logs in-between to confirm this_).

Without either of the account lines, no changes are present for detection. With the `setup email add` with `docker exec` approach only, we get 1 change detected, even when using it to add two accounts (though it's possible change detection could occur slightly inbetween those two updates, sometimes resulting in two change events.

---

Here is the logs with a simple to reproduce test:

```bash
wait_for_finished_setup_in_container mail
sleep 60

# The problem line that creates two change events:
docker run --rm -t "${NAME}" /bin/sh -c 'echo "# comment"' >> "${PRIVATE_CONFIG}/postfix-accounts.cf"

wait_until_change_detection_event_completes mail
sleep 60
```

```
docker exec mail sh -c "cat /var/log/supervisor/changedetector.log"

[  DEBUG  ]  2022-10-09 04:43:50  Starting changedetector
[  DEBUG  ]  2022-10-09 04:44:00  Changedetector is ready
---------------------------------------------------------------------------------------------------------------------------------------------------------
[   INF   ]  2022-10-09 04:44:00  Change detected
---------------------------------------------------------------------------------------------------------------------------------------------------------
user1@localhost.localdomain|{SHA512-CRYPT}$6$PPeLajmedrqhKiTx$z.XP2qjPS196IhAZDsQ/JI.pm2HKCBaDZZG/sQ9qbSlpEbZoslwaN7WvBbXiCB662yTWsBjiZZTg613aZgVdj1
user2@otherdomain.tld|{SHA512-CRYPT}$6$yhZeuyPYOpdLiNUi$3zotjjHCLqa2.s4eXdrJPDyF1Oh6D8G49wRWbMfGq6HaI.IMd48KcCgJ6cg2koD2AFjs9sYWlDvDyDrR8LGz00
user3@localhost.localdomain|{SHA512-CRYPT}$6$xVgbbov7CMgZbSg2$KQyxDFh36Sh1ha8ZyoU9CCr/DwjcJnIaWA.Sooya7gCmg9Yo4/UeHS5tPnBaKXOKyed3jkECnIatqknUr3zyX/|userdb_mail=mbox:~/mail:INBOX=~/inbox
# this is a test comment, please don't delete me :'(
           # this is also a test comment, :O
# comment
---------------------------------------------------------------------------------------------------------------------------------------------------------
[  DEBUG  ]  Creating master user 'masterusername'
[  DEBUG  ]  Creating user 'user1' for domain 'localhost.localdomain'
[  DEBUG  ]  Creating user 'user2' for domain 'otherdomain.tld'
[  DEBUG  ]  Creating user 'user3' for domain 'localhost.localdomain' with attributes 'userdb_mail=mbox:~/mail:INBOX=~/inbox'
[  DEBUG  ]  Adding alias 'alias1@localhost.localdomain' for user 'user1@localhost.localdomain' to Dovecot's userdb
[  DEBUG  ]  Alias 'alias2@localhost.localdomain' is non-local (or mapped to a non-existing account) and will not be added to Dovecot's userdb
[  DEBUG  ]  Adding alias '@localdomain2.com' for user 'user1@localhost.localdomain' to Dovecot's userdb
[  DEBUG  ]  2022-10-09 04:44:01  Restarting services due to detected changes
postfix: stopped
postfix: started
dovecot: stopped
dovecot: started
[  DEBUG  ]  2022-10-09 04:44:06  Completed handling of detected change
---------------------------------------------------------------------------------------------------------------------------------------------------------
[   INF   ]  2022-10-09 04:44:08  Change detected
---------------------------------------------------------------------------------------------------------------------------------------------------------
user1@localhost.localdomain|{SHA512-CRYPT}$6$PPeLajmedrqhKiTx$z.XP2qjPS196IhAZDsQ/JI.pm2HKCBaDZZG/sQ9qbSlpEbZoslwaN7WvBbXiCB662yTWsBjiZZTg613aZgVdj1
user2@otherdomain.tld|{SHA512-CRYPT}$6$yhZeuyPYOpdLiNUi$3zotjjHCLqa2.s4eXdrJPDyF1Oh6D8G49wRWbMfGq6HaI.IMd48KcCgJ6cg2koD2AFjs9sYWlDvDyDrR8LGz00
user3@localhost.localdomain|{SHA512-CRYPT}$6$xVgbbov7CMgZbSg2$KQyxDFh36Sh1ha8ZyoU9CCr/DwjcJnIaWA.Sooya7gCmg9Yo4/UeHS5tPnBaKXOKyed3jkECnIatqknUr3zyX/|userdb_mail=mbox:~/mail:INBOX=~/inbox
# this is a test comment, please don't delete me :'(
           # this is also a test comment, :O
# comment
---------------------------------------------------------------------------------------------------------------------------------------------------------
[  DEBUG  ]  Creating master user 'masterusername'
[  DEBUG  ]  Creating user 'user1' for domain 'localhost.localdomain'
[  DEBUG  ]  Creating user 'user2' for domain 'otherdomain.tld'
[  DEBUG  ]  Creating user 'user3' for domain 'localhost.localdomain' with attributes 'userdb_mail=mbox:~/mail:INBOX=~/inbox'
[  DEBUG  ]  Adding alias 'alias1@localhost.localdomain' for user 'user1@localhost.localdomain' to Dovecot's userdb
[  DEBUG  ]  Alias 'alias2@localhost.localdomain' is non-local (or mapped to a non-existing account) and will not be added to Dovecot's userdb
[  DEBUG  ]  Adding alias '@localdomain2.com' for user 'user1@localhost.localdomain' to Dovecot's userdb
[  DEBUG  ]  2022-10-09 04:44:08  Restarting services due to detected changes
postfix: stopped
postfix: started
dovecot: stopped
dovecot: started
[  DEBUG  ]  2022-10-09 04:44:14  Completed handling of detected change
```

I added the `---` to better section the log, should make it easier to spot the change detection and the additional `postfix-accounts.cf` content that I added to `check-for-changes.sh` to see if the content was different at all.

The detected checksums are different (_as they should be to detect the event_), so I'm not sure what is being changed. Possibly due to the 2 second loop window, the checksums are generated while the file is still writing the update, so the 2nd time around it would be a different checksum?... I ruled that out by increasing the `check-for-changes.sh` loop time to 30 seconds, and we still get 2 change events, just with a larger time apart. So the file is likely being manipulated during the change event processing.

Meanwhile, neither of these reproduce the issue, both only cause 1 change event when the only line:

```bash
docker exec -it mail /bin/sh -c 'echo "# comment" >> "/tmp/docker-mailserver/postfix-accounts.cf"'
echo "# comment" >> "${PRIVATE_CONFIG}/postfix-accounts.cf"
```

Used `cmp -b file-before file-after` to detect diff by bytes at the start and end of a change event. It revealed that `^M` (CRLF: `\r\n`) and `^J` (LF: `\n`) are the difference... line endings :grimacing: well that was annoying :laughing: 

</details>

---

## Resolved :tada: 

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/test/tests.bats#L47

This particular way of adding an account causes two change events as it inserts a CRLF instead of a LF, and when `postfix-accounts.cf` is a change event trigger it will run the `helpers/accounts.sh` script that has this line converting CRLF into LF:

https://github.com/docker-mailserver/docker-mailserver/blob/ff969509f8b9c49e399327887c9a29f6fd664b80/target/scripts/helpers/accounts.sh#L25

There is similar line duplicated down at the bottom of the file for the master accounts feature, and likewise `helpers/aliases.sh` has a similar method that can modify a config file.

I don't think we need this processing, if we do we could probably it to a temporary file rather than modify the existing config.

Reply:
[I chose just to fix only `tests.bats`](https://github.com/docker-mailserver/docker-mailserver/pull/2820) for now instead.

I don't have time to look into changing the helper scripts, and rather address that by adopting something like rust to do the processing instead :sweat_smile:

Reply:
Marking this as resolved :)